### PR TITLE
feat: horizon

### PR DESCRIPTION
**Pull request by ashitakah**
# 🤖 Linear

Closes GRT-17

### CODE REVIEW

File: src/contracts/HorizonAccountingExtension.sol [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.26;
+
+import {IHorizonAccountingExtension, IHorizonStaking} from 'interfaces/IHorizonAccountingExtension.sol';
+
+contract HorizonAccountingExtension is IHorizonAccountingExtension {
+  /// @inheritdoc IHorizonAccountingExtension
+  uint256 public immutable MIN_THAWING_PERIOD = 30 days;
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:8]: Lets set it in the constructor

File: src/contracts/HorizonAccountingExtension.sol [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  mapping(address _user => uint256 _bonded) public totalBonded;
+
+  constructor(IHorizonStaking _horizonStaking, address _prophet) {
+    horizonStaking = _horizonStaking;
+    prophet = _prophet;
+  }
+
+  /// @inheritdoc IHorizonAccountingExtension
+  function bondedAction(uint256 _bondAmount) external {
+    IHorizonStaking.Provision memory _provisionData = horizonStaking.getProvision(msg.sender, prophet);
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:26]:
```suggestion
    IHorizonStaking.Provision memory _provisionData = horizonStaking.getProvision(msg.sender, address(this));
```

The verifier will be this contract

File: src/contracts/HorizonAccountingExtension.sol [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    if (_provisionData.thawingPeriod != MIN_THAWING_PERIOD) revert HorizonAccountingExtension_InvalidThawingPeriod();
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:28]: Add a todo to review checking the max verifier cut too

File: src/contracts/HorizonAccountingExtension.sol [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    if (totalBonded[msg.sender] > _provisionData.tokens + _provisionData.tokensThawing) {
+      revert HorizonAccountingExtension_InsufficientBondedTokens();
+    }
+
+    emit Bonded(msg.sender, _bondAmount);
+  }
+
+  /// @inheritdoc IHorizonAccountingExtension
+  function finalize(uint256 _bondAmount) external {
+    if (_bondAmount > totalBonded[msg.sender]) revert HorizonAccountingExtension_InsufficientBondedTokens();
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:42]: This check is not needed as it should never happen. Just let it revert with the -=

File: src/contracts/HorizonAccountingExtension.sol [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.26;
+
+import {IHorizonAccountingExtension, IHorizonStaking} from 'interfaces/IHorizonAccountingExtension.sol';
+
+contract HorizonAccountingExtension is IHorizonAccountingExtension {
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:6]: Use the accounting extension interface

File: src/contracts/HorizonAccountingExtension.sol [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  /// @inheritdoc IHorizonAccountingExtension
+  uint256 public immutable MIN_THAWING_PERIOD = 30 days;
+
+  /// @inheritdoc IHorizonAccountingExtension
+  IHorizonStaking public horizonStaking;
+
+  /// @inheritdoc IHorizonAccountingExtension
+  address public prophet;
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:14]: immutable

File: src/contracts/HorizonAccountingExtension.sol [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+pragma solidity 0.8.26;
+
+import {IHorizonAccountingExtension, IHorizonStaking} from 'interfaces/IHorizonAccountingExtension.sol';
+
+contract HorizonAccountingExtension is IHorizonAccountingExtension {
+  /// @inheritdoc IHorizonAccountingExtension
+  uint256 public immutable MIN_THAWING_PERIOD = 30 days;
+
+  /// @inheritdoc IHorizonAccountingExtension
+  IHorizonStaking public horizonStaking;
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:11]: immutable

File: src/contracts/HorizonAccountingExtension.sol [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    emit Bonded(msg.sender, _bondAmount);
+  }
+
+  /// @inheritdoc IHorizonAccountingExtension
+  function finalize(uint256 _bondAmount) external {
+    if (_bondAmount > totalBonded[msg.sender]) revert HorizonAccountingExtension_InsufficientBondedTokens();
+    totalBonded[msg.sender] -= _bondAmount;
+
+    emit Finalized(msg.sender, _bondAmount);
+  }
```
Comment (0xShaito) [inline src/contracts/HorizonAccountingExtension.sol:46]: Missing the slash action

**Review by 0xShaito — CHANGES_REQUESTED** [review]

File: test/unit/HorizonAccountingExtension.t.sol [hunk 9]
```diff
@@ -0,0 +1,654 @@
...
+    );
+
+    vm.expectEmit();
+    emit EscalationRewardClaimed(_requestId, _disputeId, _pledger, _amount * _pledgesAgainstDispute);
+
+    horizonAccountingExtension.claimEscalationReward(_disputeId, _pledger);
+  }
+}
+
+// contract HorizonAccountingExtension_Unit_BondedAction is HorizonAccountingExtension_Unit_BaseTest {
```
Comment (0xJabberwock) [inline test/unit/HorizonAccountingExtension.t.sol:526]: Unit tests for many functions are lacking, but that shouldn't be a burden considering that they could potentially be inherited from Prophet, entirely.

**Review by 0xJabberwock — CHANGES_REQUESTED** [review]

File: src/interfaces/IHorizonAccountingExtension.sol [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+  /*///////////////////////////////////////////////////////////////
+                              LOGIC
+  //////////////////////////////////////////////////////////////*/
+
+  /**
+   * @notice Returns the approved modules for bonding tokens
+   * @param _user The address of the user
+   * @return _approvedModules The approved modules for bonding tokens
+   */
+  function approvedModules(address _user) external view returns (address[] memory _approvedModules);
```
[the remark this answers is not in the document]
Comment (ashitakah) [reply src/interfaces/IHorizonAccountingExtension.sol:251]: 👍

File: test/utils/Helpers.sol [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+   * @notice Ensures that a fuzzed address can be used for deployment and calls
+   *
+   * @param _address The address to check
+   */
+  function _assumeFuzzable(address _address) internal view {
+    assumeNotForgeAddress(_address);
+    assumeNotZeroAddress(_address);
+    assumeNotPrecompile(_address, block.chainid); // using Optimism chaind id for precompiles filtering
```
[the remark this answers is not in the document]
Comment (ashitakah) [reply test/utils/Helpers.sol:87]: 👍

File: test/utils/Helpers.sol [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+  IOracle.Response public mockResponse =
+    IOracle.Response({proposer: proposer, requestId: _mockRequestId, response: bytes('')});
+
+  bytes32 internal _mockResponseId = keccak256(abi.encode(mockResponse));
+
+  IOracle.Dispute public mockDispute =
+    IOracle.Dispute({disputer: disputer, responseId: _mockResponseId, proposer: proposer, requestId: _mockRequestId});
+
+  bytes32 internal _mockDisputeId = keccak256(abi.encode(mockDispute));
```
[the remark this answers is not in the document]
Comment (ashitakah) [reply test/utils/Helpers.sol:38]: 👍

**Pull request closed without merging**